### ISSUE TITLE
Can't replicate training improvement using mixed loss

### ISSUE THREAD

**Issue by mjacar** [mjacar/tensorflow-dcn-plus#1]
As the title says, I'm unable to get any improvement in training using the mixed loss instead of just the cross-entropy loss. That's a bummer considering the fact that it seems I'm able to match the published results of the strictly cross-entropy version reasonably well. I can't get the mixed loss version to do any better than about 15 points worse than the cross entropy version. There may be a bug in the implementation of the RL loss, but the RL loss decreases as expected when using the mixed loss (and, as expected, does not decrease when using the cross entropy loss) and the model performance doesn't totally diverge either. I don't really know what to make of that. :man_shrugging:

**Comment by mehdimashayekhi** [comment]
thanks for sharing this, I have some questions along the same lines and appreciate your response.  
1) why there is tf.stop_gradient here: https://github.com/mjacar/tensorflow-dcn-plus/blob/9358c85fb3b66e67d770b501655e89c62507a326/loss.py#L34

2) your calculation of surr_loss seems correct (https://github.com/mjacar/tensorflow-dcn-plus/blob/9358c85fb3b66e67d770b501655e89c62507a326/loss.py#L90). I can just understand why you are doing this ```tf.stop_gradient(loss - surr_loss)```. I couldn't understand the comments you provide there as well regarding this. Would you mind explain this further?

**Comment by mjacar** [comment]
1. I don't think there's any good reason for that `tf.stop_gradient`. I think it can be safely removed.
2. The surrogate loss is not the value that we actually care about; the loss is. The problem is that the loss is not differentiable, so we invent this concept of the surrogate loss which is just some function that's differentiable whose gradient is a good estimate for the gradient of the actual loss (in expectation). The surrogate loss in and of itself isn't a meaningful quantity. Therefore, if we just pass `surr_loss` through there, then `theta_rl` will get nonsensical updates and the model will diverge almost immediately. Does that make sense?

**Comment by mehdimashayekhi** [comment]
thanks much for your reply. I was wondering if you can give more theoretical justification. based on my understanding policy gradient loss is this ```J(θ)=∑dπ(s)Vπ(s)=∑dπ(s)∑πθ(a|s)Qπ(s,a)``` (http://www0.cs.ucl.ac.uk/staff/d.silver/web/Teaching_files/pg.pdf), the difference here between cross entropy and RL loss, is for RL, we multiply likelihood by reward/return (i.e., Q in formula). It is still unclear to me why you did calculate the loss the way you did, is there any paper or other implementation you can guide me to have done similar approach like you did here?

**Comment by mjacar** [comment]
Yeah, it's kind of subtle and it's one of the "Aha!" moments I had when building this. You don't typically see people actually computing an estimate for the RL loss because they don't necessarily need to. They just compute the surrogate loss and SGD will just take care of everything for them; the actual value of the RL loss isn't important in the update step. The distinction in this case is the use of the mixed objective loss where the actual value of the RL loss plays a role in properly executing a correct update step.

**Comment by mehdimashayekhi** [comment]
@mjacar excellent point, thanks again
